File: tools/binding_compliance/conformance/families/file_generation.py
"""File-generation facts require actual creation flags and unchanged existing bytes."""

from collections.abc import Mapping
from typing import Any

from ..coverage import CoveragePredicate, FamilyCoveragePolicy
from .file_operations import _files


def _observed(observation: Mapping[str, Any]) -> bool:
    """Require complete durable inventory and idempotent object/free-function calls."""
    if set(observation) != {
        "paths",
        "before",
        "generated",
        "existing",
        "standalone",
        "files",
    }:
        return False
    paths = observation["paths"]
    if (
        not isinstance(paths, list)
        or len(paths) != 2
        or paths[0] != "CLASSIC Ignore.yaml"
        or paths[1]
        not in {
            "CLASSIC Data/CLASSIC Fallout4 Local.yaml",
            "CLASSIC Data/CLASSIC Fallout4VR Local.yaml",
        }
    ):
        return False
    if not _files(observation["before"]) or not _files(observation["files"]):
        return False
    before = {item["path"]: item["content"] for item in observation["before"]}
    after = {item["path"]: item["content"] for item in observation["files"]}
    return (
        observation["generated"] == [path not in before for path in paths]
        and all(type(value) is bool for value in observation["generated"])
        and observation["existing"] == observation["standalone"] == [False, False]
        and set(after) == set(before) | set(paths)
        and all(after[path] == content for path, content in before.items())
    )


FILE_GENERATION_COVERAGE_POLICY = FamilyCoveragePolicy(
    "file-generation",
    (
        CoveragePredicate(
            "generated-and-preserved",
            "file-generation.generate",
            "file-generation.generate",
            "generated-files",
            (
                "FileGenerator",
                "FileGeneratorConfig",
                "generate_ignore_file",
                "generate_local_yaml",
            ),
            _observed,
            runtime_operations=(
                None,
                "__init__",
                "config",
                "generate_all_files_async",
                "generate_ignore_file_async",
                "generate_local_yaml_async",
                "ignore_file_path",
                "local_yaml_path",
                "generateIgnoreFile",
                "generateLocalYaml",
            ),
        ),
    ),
)

File: business-logic/classic-scanlog-core/tests/semantic_conformance/file_generation.rs
//! Native generation observations in an isolated process-relative working root.

use super::file_operations::{destination, files};
use super::{RunnerResult, invalid, text};
use classic_file_io_core::generation::{
    FileGenerator, FileGeneratorConfig, generate_ignore_file, generate_local_yaml,
};
use serde_json::{Value, json};
use std::fs;

/// Executes every native generation API and restores cwd after all awaited work.
pub(super) fn observe(fixture: &Value) -> RunnerResult<Value> {
    let game = text(&fixture["game"])?;
    if !matches!(game.as_str(), "Fallout4" | "Fallout4VR") {
        return Err(invalid("invalid generator game").into());
    }
    let temporary = tempfile::tempdir()?;
    let root = temporary.path();
    for (path, content) in fixture["files"]
        .as_object()
        .ok_or_else(|| invalid("files must be an object"))?
    {
        let target = destination(root, path)?;
        fs::create_dir_all(
            target
                .parent()
                .ok_or_else(|| invalid("file needs parent"))?,
        )?;
        fs::write(target, text(content)?.as_bytes())?;
    }
    let previous = std::env::current_dir()?;
    // The dedicated receipt process runs scenarios serially; native generators
    // use relative paths, so every future is joined before restoring the cwd.
    std::env::set_current_dir(root)?;
    let result = (|| -> RunnerResult<Value> {
        let config = FileGeneratorConfig::new(
            text(&fixture["ignore"])?,
            text(&fixture["local"])?,
            game.clone(),
        );
        let generator = FileGenerator::new(config.clone());
        let copied = generator.config();
        if copied.ignore_file_content != config.ignore_file_content
            || copied.local_yaml_content != config.local_yaml_content
            || copied.game_name != config.game_name
        {
            return Err(invalid("generator config accessor changed state").into());
        }
        let mut observation = json!({"paths":[generator.ignore_file_path().to_string_lossy().replace('\\',"/"),generator.local_yaml_path().to_string_lossy().replace('\\',"/")],"before":files(root)?});
        let runtime = classic_shared_core::get_runtime();
        observation["generated"] = json!(runtime.block_on(generator.generate_all_files_async())?);
        observation["existing"] = json!([
            runtime.block_on(generator.generate_ignore_file_async())?,
            runtime.block_on(generator.generate_local_yaml_async())?
        ]);
        observation["standalone"] = json!([
            runtime.block_on(generate_ignore_file("must not replace"))?,
            runtime.block_on(generate_local_yaml("must not replace", game))?
        ]);
        observation["files"] = files(root)?;
        Ok(observation)
    })();
    std::env::set_current_dir(previous)?;
    result
}
